EIPS/eip-8032.md

```diff
@@ -29,59 +31,153 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
+
+- `23,000,000 × 6 = 138,000,000` bytes ≈ **138 MB**
+
+This is a conservative upper bound. In practice, most contracts have far fewer slots (`L ≤ 1–3`), so a typical addition is 2–4 bytes per contract. Therefore, the net impact on state size is negligible.
 
 ## Rationale
 
 The intent is to create friction in when growing the state size of a contract, thus limiting the number of such contracts. Going over the limit, some contract developers might want to use another contract to start fresh, which comes at the cost of paying for contract creation, and for any call into the previous instance of the contract.
 
-The depth of the tree is chosen over the leaf count, because the storage tree is balanced, owing to the fact that its keys are hashed. Therefore, checking the maximum depth of a write is a stochastic process that on average, will return the size of the tree.
+`ACTIVATION_THRESHOLD` is chosen to not penalize the contracts that are large, but do provide useful value. The idea is to disincentivize spam contracts who grow larger than useful contracts, that are legitimately big due to the fact they bring value to their users and the wider ecosystem. This means that this constant could be increased as the state grows.
 
-The increase by one at each block is meant to mitigate the effects of an unbalanced tree in case of an "unlucky" contract having a very deep branch while being mostly empty: writes then become more expensive over time. On the other hand, spam accounts that see frequent writes, will see the price increase kick in after just a few blocks.
+### Comparison with depth-based pricing
 
-The depth is chosen as a proxy for the total leaf count, as there is a logarithmic relationship between the two in a balanced tree.
+The largest contract observed (XEN) sits around depth 9, meaning its relevant storage keys share a 9-nibble (~4.5 bytes) prefix along their MPT paths. Because the storage key into the trie is `keccak256(slot_key)`, an attacker can easily calculate the inputs offchain to find hashes and populate all contracts such that they pay the same cost as XEN.
 
-`ACTIVATION_THRESHOLD` is chosen to not penalize the contracts that are large, but do provide useful value. The idea is to disincentivize spam contracts who grow larger than useful contracts, that are legitimately big due to the fact they bring value to their users and the wider ecosystem. This means that this constant could be increased as the state grows.
+### Why is a transition needed?
+
+If counting began only at `FORK_TIMESTAMP`, every existing contract would start with an implicit `storage_count = 0` and accrue counts solely from *new* writes. This would misprice precisely the workloads we intend to regulate: large, pre-fork contracts would continue paying unscaled base costs until they happen to rewrite a substantial portion of their historical slots. Gas pricing would under-reflect their actual storage footprint.
```

Comment:
I don't quite agree with that btw, yes it would misprice but the spammy ones would quickly be caught spamming. Nonetheless, I agree that this would not penalize growth for a while and as a result it should be done properly.